nexus/reconfigurator/planning/src/planner.rs

```diff
@@ -2180,4 +2216,631 @@ mod test {
+            log.clone(),
+            &blueprint1,
+            &input,
+            "test_blueprint2",
+            &collection,
+        )
+        .expect("created planner")
+        .with_rng_seed((TEST_NAME, "bp2"))
+        .plan()
+        .expect("plan");
+
+        // We should see zones for 3 clickhouse keepers, and 2 servers created
+        let active_zones: Vec<_> = blueprint2
+            .all_omicron_zones(BlueprintZoneFilter::ShouldBeRunning)
+            .map(|(_, z)| z.clone())
+            .collect();
+
+        let keeper_zone_ids: BTreeSet<_> = active_zones
+            .iter()
+            .filter(|z| z.zone_type.is_clickhouse_keeper())
+            .map(|z| z.id)
+            .collect();
+        let server_zone_ids: BTreeSet<_> = active_zones
+            .iter()
+            .filter(|z| z.zone_type.is_clickhouse_server())
+            .map(|z| z.id)
+            .collect();
+
+        assert_eq!(keeper_zone_ids.len(), target_keepers);
+        assert_eq!(server_zone_ids.len(), target_servers);
+
+        // We should be attempting to allocate all servers and keepers since
+        // this the initial configuration
+        {
+            let clickhouse_cluster_config =
+                blueprint2.clickhouse_cluster_config.as_ref().unwrap();
+            assert_eq!(clickhouse_cluster_config.generation, 2.into());
+            assert_eq!(
+                clickhouse_cluster_config.max_used_keeper_id,
+                (target_keepers as u64).into()
+            );
+            assert_eq!(
+                clickhouse_cluster_config.max_used_server_id,
+                (target_servers as u64).into()
+            );
+            assert_eq!(clickhouse_cluster_config.keepers.len(), target_keepers);
+            assert_eq!(clickhouse_cluster_config.servers.len(), target_servers);
+
+            // Ensure that the added keepers are in server zones
+            for zone_id in clickhouse_cluster_config.keepers.keys() {
+                assert!(keeper_zone_ids.contains(zone_id));
+            }
+
+            // Ensure that the added servers are in server zones
+            for zone_id in clickhouse_cluster_config.servers.keys() {
+                assert!(server_zone_ids.contains(zone_id));
+            }
+        }
+
+        // Planning again without changing inventory should result in the same
+        // state
+        let blueprint3 = Planner::new_based_on(
+            log.clone(),
+            &blueprint2,
+            &input,
+            "test_blueprint3",
+            &collection,
+        )
+        .expect("created planner")
+        .with_rng_seed((TEST_NAME, "bp3"))
+        .plan()
+        .expect("plan");
+
+        assert_eq!(
+            blueprint2.clickhouse_cluster_config,
+            blueprint3.clickhouse_cluster_config
+        );
+
+        // Updating the inventory to reflect the keepers
+        // should result in the same state, except for the
+        // `highest_seen_keeper_leader_committed_log_index`
```

Comment:
Just making sure I understand this - does this mean we expect the blueprint contents to change regularly in a way that doesn't really matter while the system is in a steady state? I think we kind of already have something like that in `{internal,external}_dns_version` (which can change in response to silo creation or deletion, for example, even in a steady state). I don't think there's anything wrong here; I want to understand for when we start talking about automating blueprint generation and deciding when to set a new target (e.g., if they _only_ thing that's changed is from blueprint A to blueprint B `highest_seen_keeper_leader_committed_log_index`, then we'd say there's no need to advance the target from A to B?).

Reply:
> does this mean we expect the blueprint contents to change regularly in a way that doesn't really matter while the system is in a steady state?

Yep. The keeper will continue to commit entries to the raft log and inventory will continue to pick up the update.

> I want to understand for when we start talking about automating blueprint generation and deciding when to set a new target (e.g., if they only thing that's changed is from blueprint A to blueprint B highest_seen_keeper_leader_committed_log_index, then we'd say there's no need to advance the target from A to B?).

Yes, this log index is mainly used as an inventory generation number that maps to the configuration. It is solely for allowing us to know when a changed configuration is newer than another configuration since we pull configurations from multiple nodes and different nodes can be online or offline at a given time. If the configuration in the inventory isn't differen than what's in the blueprint then we don't need to generate a new blueprint.